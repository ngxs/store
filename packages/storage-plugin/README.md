# @ngxs/storage-plugin

Storage plugin for NGXS. See [repo](https://github.com/ngxs/store) for more info.
