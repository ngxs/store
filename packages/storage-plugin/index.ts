/**
 * The public api for consumers of @ngxs/storage-plugin
 */
export * from './src/public_api';
