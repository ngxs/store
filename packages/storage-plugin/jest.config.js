module.exports = {
  displayName: 'storage-plugin',
  preset: '../../jest.preset.js',
  setupFilesAfterEnv: ['<rootDir>/test-setup.ts'],
  globals: {},
  coverageDirectory: '../../coverage/packages/storage-plugin',
  transform: {
    '^.+\\.(ts|mjs|js|html)$': [
      'jest-preset-angular',
      {
        isolatedModules: true,
        tsconfig: '<rootDir>/tsconfig.spec.json',
        stringifyContentPathRegex: '\\.(html|svg)$'
      }
    ]
  },
  snapshotSerializers: [
    'jest-preset-angular/build/serializers/no-ng-attributes',
    'jest-preset-angular/build/serializers/ng-snapshot',
    'jest-preset-angular/build/serializers/html-comment'
  ]
};
