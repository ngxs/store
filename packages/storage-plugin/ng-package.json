{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../builds/storage-plugin",
  "lib": {
    "flatModuleFile": "ngxs-storage-plugin",
    "entryFile": "index.ts",
    "umdModuleIds": {
      "@ngxs/store": "ngxs-store"
    }
  }
}
