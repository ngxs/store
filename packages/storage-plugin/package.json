{
  "name": "@ngxs/storage-plugin",
  "description": "extendable storage plugin for @ngxs/store",
  "version": "0.0.0",
  "sideEffects": false,
  "peerDependencies": {
    "@ngxs/store": "^0.0.0",
    "@angular/core": ">=12.0.0 <18.0.0",
    "rxjs": ">=6.5.5",
    "ts-morph": "21.0.1"
  },
  "schematics": "./schematics/collection.json"
}
