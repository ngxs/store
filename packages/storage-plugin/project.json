{
  "name": "storage-plugin",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/storage-plugin",
  "projectType": "library",
  "targets": {
    "build-package": {
      "executor": "@angular-devkit/build-angular:ng-packagr",
      "outputs": ["{workspaceRoot}/@ngxs/storage-plugin"],
      "options": {
        "tsConfig": "tsconfig.build.json",
        "project": "packages/storage-plugin/ng-package.json"
      }
    },
    "build": {
      "executor": "nx:run-commands",
      "options": {
        "command": "node tools/build-schematics.mjs --projectRoot=packages/storage-plugin --distPath=@ngxs/storage-plugin"
      },
      "dependsOn": ["build-package"]
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "packages/storage-plugin/jest.config.js"
      }
    },
    "lint": {
      "executor": "@nx/eslint:lint",
      "options": {
        "lintFilePatterns": [
          "packages/storage-plugin/**/*.ts",
          "packages/storage-plugin/**/*.html"
        ]
      }
    }
  }
}
