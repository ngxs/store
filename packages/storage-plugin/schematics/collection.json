{
  "extends": ["@schematics/angular"],
  "$schema": "../node_modules/@angular-devkit/schematics/collection-schema.json",
  "schematics": {
    "keys-migration": {
      "factory": "./src/ng-generate/index#migrateKeys",
      "description": "Migrates from key to keys",
      "aliases": ["ngxs-keys-migration"],
      "schema": "./src/ng-generate/schema.json"
    }
  }
}
