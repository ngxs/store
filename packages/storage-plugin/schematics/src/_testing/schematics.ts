import { SchematicTestRunner, UnitTestTree } from '@angular-devkit/schematics/testing';
import { workspaceRoot } from '@nrwl/devkit';
import { Schema as ApplicationOptions } from '@schematics/angular/application/schema';
import { Schema as WorkspaceOptions } from '@schematics/angular/workspace/schema';
import * as path from 'path';

const angularSchematicRunner = new SchematicTestRunner(
  '@schematics/angular',
  path.join(workspaceRoot, 'node_modules/@schematics/angular/collection.json')
);

const defaultWorkspaceOptions: WorkspaceOptions = {
  name: 'workspace',
  newProjectRoot: 'projects',
  version: '1.0.0'
};

const defaultAppOptions: ApplicationOptions = {
  name: 'foo',
  inlineStyle: false,
  inlineTemplate: false,
  routing: true,
  skipTests: false,
  skipPackageJson: false
};

export async function createWorkspace(
  standalone = false,
  schematicRunner = angularSchematicRunner,
  workspaceOptions = defaultWorkspaceOptions,
  appOptions = defaultAppOptions
) {
  let appTree: UnitTestTree = await schematicRunner.runSchematic(
    'workspace',
    workspaceOptions
  );
  appTree = await schematicRunner.runSchematic(
    'application',
    {
      ...appOptions,
      // Specify explicitly because it's truthy by default since Angular 17.
      standalone
    },
    appTree
  );

  return appTree;
}
