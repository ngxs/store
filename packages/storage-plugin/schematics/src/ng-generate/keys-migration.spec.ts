import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { workspaceRoot } from '@nrwl/devkit';
import { join } from 'path';
import { createWorkspace } from '../_testing';

describe('Storage Plugin Migration', () => {
  const ngxsSchematicRunner = new SchematicTestRunner(
    '@ngxs/storage-plugin/schematics',
    join(workspaceRoot, 'packages/storage-plugin/schematics/collection.json')
  );

  const testSetup = async () => {
    const appTree = await createWorkspace();
    appTree.overwrite('package.json', `{"dependencies": {"@ngxs/store": "3.8.2"}}`);

    return { appTree };
  };

  it('migrate empty forRoot', async () => {
    const { appTree } = await testSetup();

    const newContent = `
    import { NgxsStoragePluginModule } from '@ngxs/storage-plugin';
    import { NgModule } from '@angular/core';
    import { BrowserModule } from '@angular/platform-browser';

    import { AppRoutingModule } from './app-routing.module';
    import { AppComponent } from './app.component';

      @NgModule({
        declarations: [
          AppComponent
        ],
        imports: [
          BrowserModule,
          AppRoutingModule,
          NgxsStoragePluginModule.forRoot()
        ],
        bootstrap: [AppComponent]
      })
      export class AppModule { }
    `;

    appTree.overwrite('/projects/foo/src/app/app.module.ts', newContent);

    const tree = await ngxsSchematicRunner.runSchematic('ngxs-keys-migration', {}, appTree);

    const contentUpdate = tree!.readContent('/projects/foo/src/app/app.module.ts');
    expect(contentUpdate).toMatchSnapshot();
  });

  it('migrate forRoot that is missing the "key" property', async () => {
    const { appTree } = await testSetup();

    const newContent = `
    import { NgxsStoragePluginModule } from '@ngxs/storage-plugin';
    import { NgModule } from '@angular/core';
    import { BrowserModule } from '@angular/platform-browser';

    import { AppRoutingModule } from './app-routing.module';
    import { AppComponent } from './app.component';

      @NgModule({
        declarations: [
          AppComponent
        ],
        imports: [
          BrowserModule,
          AppRoutingModule,
          NgxsStoragePluginModule.forRoot({proa: 'valuea', prob: 'valueb'})
        ],
        bootstrap: [AppComponent]
      })
      export class AppModule { }
    `;

    appTree.overwrite('/projects/foo/src/app/app.module.ts', newContent);

    const tree = await ngxsSchematicRunner.runSchematic('ngxs-keys-migration', {}, appTree);

    const contentUpdate = tree!.readContent('/projects/foo/src/app/app.module.ts');
    expect(contentUpdate).toMatchSnapshot();
  });

  it('migrate forRoot that has the "key" property with a value as a string literal', async () => {
    const { appTree } = await testSetup();

    const newContent = `
    import { NgxsStoragePluginModule } from '@ngxs/storage-plugin';
    import { NgModule } from '@angular/core';
    import { BrowserModule } from '@angular/platform-browser';

    import { AppRoutingModule } from './app-routing.module';
    import { AppComponent } from './app.component';

      @NgModule({
        declarations: [
          AppComponent
        ],
        imports: [
          BrowserModule,
          AppRoutingModule,
          NgxsStoragePluginModule.forRoot({proa: 'valuea', prob: 'valueb', key: 'keep this value'})
        ],
        bootstrap: [AppComponent]
      })
      export class AppModule { }
    `;

    appTree.overwrite('/projects/foo/src/app/app.module.ts', newContent);

    const tree = await ngxsSchematicRunner.runSchematic('ngxs-keys-migration', {}, appTree);

    const contentUpdate = tree!.readContent('/projects/foo/src/app/app.module.ts');
    expect(contentUpdate).toMatchSnapshot();
  });

  it('migrate forRoot that has the "key" property with values as an array literal', async () => {
    const { appTree } = await testSetup();

    const newContent = `
    import { NgxsStoragePluginModule } from '@ngxs/storage-plugin';
    import { NgModule } from '@angular/core';
    import { BrowserModule } from '@angular/platform-browser';

    import { AppRoutingModule } from './app-routing.module';
    import { AppComponent } from './app.component';

      @NgModule({
        declarations: [
          AppComponent
        ],
        imports: [
          BrowserModule,
          AppRoutingModule,
          NgxsStoragePluginModule.forRoot({proa: 'valuea', prob: 'valueb', key: ['keyValue', NovelsState, { key: 'novels', engine: MyEngineX } ]})
        ],
        bootstrap: [AppComponent]
      })
      export class AppModule { }
    `;

    appTree.overwrite('/projects/foo/src/app/app.module.ts', newContent);

    const tree = await ngxsSchematicRunner.runSchematic('ngxs-keys-migration', {}, appTree);

    const contentUpdate = tree!.readContent('/projects/foo/src/app/app.module.ts');
    expect(contentUpdate).toMatchSnapshot();
  });

  it('migrate from a constant array', async () => {
    const { appTree } = await testSetup();

    const newContent = `
    import { NgxsStoragePluginModule } from '@ngxs/storage-plugin';
    import { NgModule } from '@angular/core';
    import { BrowserModule } from '@angular/platform-browser';

    import { AppRoutingModule } from './app-routing.module';
    import { AppComponent } from './app.component';

    const imports = [
      BrowserModule,
      AppRoutingModule,
      NgxsStoragePluginModule.forRoot({proa: 'valuea', prob: 'valueb', key: 'keep this value'})
    ];

      @NgModule({
        declarations: [
          AppComponent
        ],
        imports,
        bootstrap: [AppComponent]
      })
      export class AppModule { }
    `;

    appTree.overwrite('/projects/foo/src/app/app.module.ts', newContent);

    const tree = await ngxsSchematicRunner.runSchematic('ngxs-keys-migration', {}, appTree);

    const contentUpdate = tree!.readContent('/projects/foo/src/app/app.module.ts');
    expect(contentUpdate).toMatchSnapshot();
  });
});
