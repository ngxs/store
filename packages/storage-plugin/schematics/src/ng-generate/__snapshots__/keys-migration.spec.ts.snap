// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storage Plugin Migration migrate empty forRoot 1`] = `
"
    import { NgxsStoragePluginModule } from '@ngxs/storage-plugin';
    import { NgModule } from '@angular/core';
    import { BrowserModule } from '@angular/platform-browser';

    import { AppRoutingModule } from './app-routing.module';
    import { AppComponent } from './app.component';

      @NgModule({
        declarations: [
          AppComponent
        ],
        imports: [
          BrowserModule,
          AppRoutingModule,
          NgxsStoragePluginModule.forRoot({ keys: '*' })
        ],
        bootstrap: [AppComponent]
      })
      export class AppModule { }
    "
`;

exports[`Storage Plugin Migration migrate forRoot that has the "key" property with a value as a string literal 1`] = `
"
    import { NgxsStoragePluginModule } from '@ngxs/storage-plugin';
    import { NgModule } from '@angular/core';
    import { BrowserModule } from '@angular/platform-browser';

    import { AppRoutingModule } from './app-routing.module';
    import { AppComponent } from './app.component';

      @NgModule({
        declarations: [
          AppComponent
        ],
        imports: [
          BrowserModule,
          AppRoutingModule,
          NgxsStoragePluginModule.forRoot({proa: 'valuea', prob: 'valueb', keys: ['keep this value']})
        ],
        bootstrap: [AppComponent]
      })
      export class AppModule { }
    "
`;

exports[`Storage Plugin Migration migrate forRoot that has the "key" property with values as an array literal 1`] = `
"
    import { NgxsStoragePluginModule } from '@ngxs/storage-plugin';
    import { NgModule } from '@angular/core';
    import { BrowserModule } from '@angular/platform-browser';

    import { AppRoutingModule } from './app-routing.module';
    import { AppComponent } from './app.component';

      @NgModule({
        declarations: [
          AppComponent
        ],
        imports: [
          BrowserModule,
          AppRoutingModule,
          NgxsStoragePluginModule.forRoot({proa: 'valuea', prob: 'valueb', keys: ['keyValue', NovelsState, { key: 'novels', engine: MyEngineX } ]})
        ],
        bootstrap: [AppComponent]
      })
      export class AppModule { }
    "
`;

exports[`Storage Plugin Migration migrate forRoot that is missing the "key" property 1`] = `
"
    import { NgxsStoragePluginModule } from '@ngxs/storage-plugin';
    import { NgModule } from '@angular/core';
    import { BrowserModule } from '@angular/platform-browser';

    import { AppRoutingModule } from './app-routing.module';
    import { AppComponent } from './app.component';

      @NgModule({
        declarations: [
          AppComponent
        ],
        imports: [
          BrowserModule,
          AppRoutingModule,
          NgxsStoragePluginModule.forRoot({proa: 'valuea', prob: 'valueb',
              keys: '*'
        })
        ],
        bootstrap: [AppComponent]
      })
      export class AppModule { }
    "
`;

exports[`Storage Plugin Migration migrate from a constant array 1`] = `
"
    import { NgxsStoragePluginModule } from '@ngxs/storage-plugin';
    import { NgModule } from '@angular/core';
    import { BrowserModule } from '@angular/platform-browser';

    import { AppRoutingModule } from './app-routing.module';
    import { AppComponent } from './app.component';

    const imports = [
      BrowserModule,
      AppRoutingModule,
      NgxsStoragePluginModule.forRoot({proa: 'valuea', prob: 'valueb', keys: ['keep this value']})
    ];

      @NgModule({
        declarations: [
          AppComponent
        ],
        imports,
        bootstrap: [AppComponent]
      })
      export class AppModule { }
    "
`;
